actgen/change_q.py

```diff
@@ -0,0 +1,242 @@
+            if arg_name == 'hyperparams':
+                continue
+            params[arg_name] = arg_value
+        for arg_name, arg_value in args.other_args:
+            utils.update_param(params, arg_name, arg_value)
+        return params
+
+    def setup(self):
+        seeding.seed(0, random, torch, np)
+        test_env = gym.make(self.params['env_name'])
+        test_env = wrap.FixedDurationHack(test_env)
+        test_env = wrap.DuplicateActions(test_env, self.params['duplicate'])
+        test_env = wrap.TorchInterface(test_env)
+        seeding.seed(1000 + self.params['seed'], gym, test_env)
+        self.test_env = test_env
+
+        assert self.params['agent'] == 'dqn'
+        self.agent = DQNAgent(test_env.observation_space, test_env.action_space, self.params)
+        # load saved model
+        if not self.params['test']:
+            self.agent.q.load(self.params['saved_model'])
+
+    def teardown(self):
+        pass
+
+    def original_q_values(self, state):
+        """
+        get the q values for all actions for a particular state
+        """
+        self.agent.q.eval()
+        q_values = self.agent.q(state.float())
+        return q_values.detach().numpy()[0]
+
+    def dqn_directed_update(self, state, action, toward_target):
+        """
+        update the dqn agent towards a target for one state.
+        specifically, update update q(state, action) towards toward_target.
+        returns the old q(state,a) and the updated q(state, a) for all a.
+        """
+        self.agent.q.train()
+        self.agent.optimizer.zero_grad()
+
+        old_q_values = self.agent.q(state.float())  # q(state, a) for all a
+        q_target = old_q_values.clone()
+        q_target[0][action] = toward_target
+
+        loss = torch.nn.functional.smooth_l1_loss(input=old_q_values, target=q_target)
+        loss.backward()
+        self.agent.optimizer.step()
+
+        self.agent.q.eval()
+        new_q_values = self.agent.q(state.float())
+
+        return old_q_values.detach().numpy()[0], new_q_values.detach().numpy()[0]
+
+    def plot(self, old_values, new_values, action_idx):
+        """
+        given a set of old and new q values, plot them in a graph to compare
+        this function plots the change in q_values in one plot for one action
+
+        old_values.shape = (env.action_space.n,)
+        new_values.shape = (num_updates, env.action_space.n)
+        """
+        possible_actions = np.arange(self.test_env.action_space.n)
+
+        # plot reference line at 0
+        plt.plot(possible_actions, np.zeros_like(possible_actions), '--k')
+
+        # plot all new values
+        for i, new_val in enumerate(new_values):
+            plt.plot(possible_actions, new_val - old_values, label='iter {}'.format(i))
+
+        plt.title('action {}'.format(action_idx))
+        plt.xlabel('{} different actions'.format(self.test_env.action_space.n))
+        plt.ylabel('q(s,a)')
```

Comment:
You can add a delta symbol to the y-label with `$\Delta$`. You may need to r-quote the string: `r'$\Delta$q(s,a)'`